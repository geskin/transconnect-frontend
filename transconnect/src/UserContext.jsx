import { createContext, useState, useEffect } from "react";

const UserContext = createContext();

export const UserProvider = ({ children }) => {
    const [currUser, setCurrUser] = useState(() => {
        const storedUser = localStorage.getItem("currUser");
        return storedUser ? JSON.parse(storedUser) : null;
    });

    // Save to localStorage whenever currUser changes
    useEffect(() => {
        if (currUser) {
            localStorage.setItem("currUser", JSON.stringify(currUser));
        } else {
            localStorage.removeItem("currUser");
        }
    }, [currUser]);

    return (
        <UserContext.Provider value={{ currUser, setCurrUser }}>
            {children}
        </UserContext.Provider>
    );
};

export default UserContext;

