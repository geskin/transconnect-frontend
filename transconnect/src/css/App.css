#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 0rem;
  text-align: center;
  box-sizing: border-box;
}

body {
  margin: 0;
  padding: 0;
  font-family: system-ui, sans-serif;
  overflow-x: hidden;
}

.logo {
  height: 6em;
  padding: 1.5em;
  will-change: filter;
  transition: filter 300ms;
}

.logo:hover {
  filter: drop-shadow(0 0 2em #646cffaa);
}

.logo.react:hover {
  filter: drop-shadow(0 0 2em #61dafbaa);
}

@keyframes logo-spin {
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
}

@media (prefers-reduced-motion: no-preference) {
  a:nth-of-type(2) .logo {
    animation: logo-spin infinite 20s linear;
  }
}

.card {
  padding: 2em;
}

.read-the-docs {
  color: #888;
}

/* Mobile responsiveness */
@media (max-width: 768px) {
  #root {
    /* padding: 1rem; */
  }

  .logo {
    height: 4em;
    padding: 1em;
  }

  .card {
    padding: 1em;
  }

  .read-the-docs {
    font-size: 0.9rem;
  }
}