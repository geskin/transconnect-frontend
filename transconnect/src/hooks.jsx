import { useState, useEffect } from "react";

const useLocalStorage = (defaultVal, key) => {
    const [val, setVal] = useState(() => {
        if (typeof window === 'undefined' || !window.localStorage) return defaultVal;

        const persistedVal = window.localStorage.getItem(key);
        // return persistedVal !== null ? JSON.parse(persistedVal) : defaultVal;

        if (!persistedVal || persistedVal === "undefined" || persistedVal === "null" || persistedVal === null || persistedVal === undefined) {
            return defaultVal;
        } else {
            try {
                return JSON.parse(persistedVal);
            } catch (error) {
                console.error("Error parsing value:", error);
                return defaultVal;
            }
        }
    });

    useEffect(() => {
        window.localStorage.setItem(key, JSON.stringify(val));
    }, [key, val]);

    return [val, setVal];
}

export { useLocalStorage };