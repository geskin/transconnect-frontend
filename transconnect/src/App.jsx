import { useContext, useEffect } from 'react'
import './css/App.css'
import { BrowserRouter } from 'react-router-dom';
import NavBar from './NavBar';
import RoutesList from './RoutesList';
import UserContext from './UserContext';
import { useLocalStorage } from './hooks';
import TransconnectApi from './api';
import { jwtDecode } from 'jwt-decode';

function App() {
  const { currUser, setCurrUser } = useContext(UserContext);
  const [token, setToken] = useLocalStorage(null, 'token');

  useEffect(() => {
    if (token) {
      TransconnectApi.token = token; // set token globally
    }
  }, [token]);

  useEffect(() => {
    if (!token) {
      setCurrUser(null);
      return;
    }

    let payload = jwtDecode(token);
    const { username, role } = payload; // Extract username and isAdmin from the token

    // Set the current user with username and isAdmin
    setCurrUser({ username, role });
  }, [token]);

  // Fetch additional user details only after curr_user is set
  useEffect(() => {
    if (!currUser || !currUser.username) return; // Wait until curr_user is fully initialized

    const fetchUser = async () => {
      try {
        const user = await TransconnectApi.getUser(currUser.username);
        setCurrUser((prevUser) => ({
          ...prevUser,
          ...user, // Merge fetched data into the current user state
        }));
      } catch (err) {
        console.error("Error fetching user info", err);
      }
    };

    fetchUser();
  }, [currUser?.username]);

  const login = async (username, password) => {
    try {
      const data = await TransconnectApi.authenticate(username, password);
      setToken(data);
    } catch (err) {
      console.error("Error authenticating user", err);
    }
  }

  const signup = async (user) => {
    try {
      const data = await TransconnectApi.register(user);
      setToken(data);
    } catch (err) {
      console.error("Error registering user", err);
    }
  }

  const logout = () => {
    setToken(null);
    setCurrUser(null);
  }

  return (
    <>
      <BrowserRouter>
        <NavBar currUser={currUser} />
        <main>
          <RoutesList signup={signup} login={login} logout={logout} />
        </main>
      </BrowserRouter>
    </>
  )
}

export default App
