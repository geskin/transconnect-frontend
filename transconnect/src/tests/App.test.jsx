import React from "react";
import { render, screen, act, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import App from "../App";
import UserContext from "../UserContext";
import TransconnectApi from "../api";
import { useLocalStorage } from "../hooks";
import { jwtDecode } from "jwt-decode";

vi.mock("../api");
vi.mock("../hooks", () => ({
    useLocalStorage: vi.fn(),
}));
vi.mock("jwt-decode", () => ({
    jwtDecode: vi.fn(() => ({ username: "testuser", role: "USER" })),
}));

beforeEach(() => {
    useLocalStorage.mockReturnValue(["test-token", vi.fn()]);
    TransconnectApi.getUser = vi.fn();
    TransconnectApi.getUser.mockResolvedValue({ email: "test@example.com" });
});

describe("App Component", () => {
    it("renders without crashing", () => {
        render(
            <UserContext.Provider value={{ currUser: null, setCurrUser: vi.fn() }}>
                <App />
            </UserContext.Provider>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <UserContext.Provider value={{ currUser: null, setCurrUser: vi.fn() }}>
                <App />
            </UserContext.Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("sets token globally when present", async () => {
        render(
            <UserContext.Provider value={{ currUser: null, setCurrUser: vi.fn() }}>
                <App />
            </UserContext.Provider>
        );
        expect(TransconnectApi.token).toBe("test-token");
    });

    it("decodes token and sets current user", async () => {
        const setCurrUserMock = vi.fn();
        jwtDecode.mockReturnValue({ username: "testuser", role: "USER" });

        render(
            <UserContext.Provider value={{ currUser: null, setCurrUser: setCurrUserMock }}>
                <App />
            </UserContext.Provider>
        );

        expect(setCurrUserMock).toHaveBeenCalledWith({ username: "testuser", role: "USER" });
    });

    it("fetches additional user data when currUser is set", async () => {
        const setCurrUserMock = vi.fn();

        TransconnectApi.getUser.mockResolvedValue({ email: "test@example.com" });

        await act(async () => {
            render(
                <UserContext.Provider value={{ currUser: { username: "testuser" }, setCurrUser: setCurrUserMock }}>
                    <App />
                </UserContext.Provider>
            );
        });

        await waitFor(() => {
            expect(TransconnectApi.getUser).toHaveBeenCalledWith("testuser");
        });

        expect(setCurrUserMock).toHaveBeenCalledTimes(2);

        expect(setCurrUserMock).toHaveBeenNthCalledWith(1, expect.objectContaining({
            username: "testuser",
            role: "USER"
        }));

        const callback = setCurrUserMock.mock.calls[1][0];
        expect(typeof callback).toBe("function");

        const updatedUser = callback({ username: "testuser", role: "USER" });
        expect(updatedUser).toEqual(expect.objectContaining({
            username: "testuser",
            email: "test@example.com",
            role: "USER"
        }));
    });
});