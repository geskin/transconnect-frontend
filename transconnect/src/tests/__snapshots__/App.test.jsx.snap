// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-lhkrlz-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-dr5mlu-MuiToolbar-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1x41qo0-MuiTypography-root"
      >
        TransConnect
      </h6>
      <div
        class="MuiBox-root css-1hq76n0"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-8chbys-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Resources
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-8chbys-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Bathrooms
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-8chbys-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Login
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-8chbys-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Signup
        </button>
      </div>
    </div>
  </header>
  <main>
    <div
      class="Homepage"
    >
      <div
        class="container text-center"
      >
        <h1
          class="mb-4 fw-bold"
        >
          TransConnect
        </h1>
        <p
          class="lead"
        >
          Connection with a cause.
        </p>
        <p>
          <button>
            <a
              class="btn btn-primary fw-bold me-3"
              data-discover="true"
              href="/login"
            >
              Log in
            </a>
          </button>
          <button>
            <a
              class="btn btn-primary fw-bold me-3"
              data-discover="true"
              href="/register"
            >
              Sign up
            </a>
          </button>
        </p>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
